resteasy/src/main/java/com/cerner/beadledom/resteasy/ModuleProcessor.java

```diff
@@ -0,0 +1,80 @@
+package com.cerner.beadledom.resteasy;
+
+import com.google.inject.Binding;
+import com.google.inject.Injector;
+
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import javax.ws.rs.ext.Provider;
+
+import org.jboss.resteasy.plugins.guice.GuiceResourceFactory;
+import org.jboss.resteasy.plugins.guice.i18n.LogMessages;
+import org.jboss.resteasy.plugins.guice.i18n.Messages;
+import org.jboss.resteasy.spi.Registry;
+import org.jboss.resteasy.spi.ResourceFactory;
+import org.jboss.resteasy.spi.ResteasyProviderFactory;
+import org.jboss.resteasy.util.GetRestful;
+
+/**
+ * Processes a given life cycle injector that contains Guice modules.
+ *
+ * <p>This class expands upon {@link org.jboss.resteasy.plugins.guice.ModuleProcessor} by enforcing
+ * restrictions on available JAX-RS {@link Provider}s to be added to {@link ResteasyProviderFactory}.
+ *
+ * @author Charan Panchagnula
+ */
+public class ModuleProcessor {
+
+  private final Registry registry;
+  private final ResteasyProviderFactory providerFactory;
+
+  /**
+   * Constructor for ModuleProcessor.
+   *
+   * @param registry {@link Registry} for adding resource implementation endpoints
+   * @param providerFactory {@link ResteasyProviderFactory} used for registering JAX-RS {@link Provider} instances
+   */
+  public ModuleProcessor(final Registry registry, final ResteasyProviderFactory providerFactory) {
+    Objects.requireNonNull(registry, "registry: null");
+    Objects.requireNonNull(providerFactory, "providerFactory: null");
+
+    this.registry = registry;
+    this.providerFactory = providerFactory;
+  }
+
+  /**
+   * Processes the injector by registering bindings for dependencies associated to each type.
+   *
+   * @param injector The {@link Injector} instance to be processed
+   */
+  public void processInjector(final Injector injector) {
+    Objects.requireNonNull(injector, "injector: null");
+
+    final List<Binding<?>> rootResourceBindings = new ArrayList<>();
+    for (final Binding<?> binding : injector.getBindings().values()) {
+      final Type type = binding.getKey().getTypeLiteral().getRawType();
+      final Class<?> beanClass = (Class) type;
+      if (GetRestful.isRootResource(beanClass)) {
+        // deferred registration
```

Comment:
Comments, especially non-javadoc ones like this, should really explain why not what. Right now this comment is more confusing than helpful to me.

Reply:
This was basically copy pastad from https://github.com/resteasy/Resteasy/blob/master/resteasy-guice/src/main/java/org/jboss/resteasy/plugins/guice/ModuleProcessor.java#L42 I can remove it

Reply:
If this is just copy pasted, do we need to include a copyright notice for Resteasy?

Reply:
Why are we just copy pasting code?

Reply:
Like i added in this comment, ModuleProcessor in resteasy-guice does not provide the functionality we want to address this issue so we decided to write our own version of this class https://github.com/charanpanchagnula/beadledom/blob/issue16_clientBindingAnnotation/resteasy/src/main/java/com/cerner/beadledom/resteasy/ModuleProcessor.java#L23

Reply:
What about changes similar to this so we aren't copying a file from another project: https://gist.github.com/johnlcox/57c451dc51d010ffb6c139de67aaa3f7?

Reply:
Functionality is more or less the same, so why are we adding bells and whistles to make it seem like its our own thing, as opposed to acknowledging that we're building upon the resteasy version (which is already mentioned in the class doc) ?

Reply:
Also, your logic seems to altogether ignore annotated bindings whereas the current logic still adds the annotated binding to rootResourceBindings list. Don't we have to register resource factory for this annotated binding?

Reply:
The functionality from Resteasy is fairly simple, and since we need to modify it in a way we cannot do via class extension I think it's simpler to write our own class/method rather than copying. 

If we are going to copy then we have a legal obligation to follow the requirements of the Apache License for the file we are copying. The StackExchange link below gives a good answer as to what we would need to do to correctly follow the license [1]. Of the steps outlined there steps 3 and 5 are going to be the ones requiring the most effort. For step 3 it will also result in muddying up the file if we have to document all of our own changes. Step 5 will require evaluating github repo of Resteasy to determine if there is a "NOTICE-like" file, and if making a decision on how/where we will keep it in our repo.



[1] https://opensource.stackexchange.com/questions/5143/how-to-mark-a-copied-apache-v2-piece-of-code

Reply:
Whether we should exclude annotated bindings of root resources or not is a good question. I'm curious what @nathanschile thinks.

Reply:
So root resources being classes that are annotated with `@Path` annotation, what would be the use case where those classes would be annotated binded? I can't think of a use case.

Reply:
https://github.com/cerner/beadledom/pull/143/commits/75cf28188e471c1af009d2dd0dfec137b1270635

there was an if-else if check for root resources and provider bindings which isn't originally there in the resteasy-guice version i stuck to if-if to align with the functionality of resteasy-guice